### PR TITLE
Fix preloader to never reset associations in case they are already loaded

### PR DESCRIPTION
**Pull request by bogdan**
### Summary

This patch fixes the issue when association is preloaded with a custom
preload scope which disposes the already preloaded target of the
association by reseting it.

When custom preload scope is used, the preloading is now performed into
a separated Hash `#records_by_owner` instead of the association.
It removes the necessity the reset the association after the preloading
is complete so that reset of the preloaded association never happens.

Preloading is still happening to the association when the preload scope
is empty.

### Motivation

The bug found can be considered small and also found analytically (by artificially finding the preload sequence that would break) - it wasn't encountered by anyone and has a pretty low chance to be found: it requires a custom association scope inside the through association and very specific sequence of arguments passed to `Relation#preload`.

IMO the implementation based on resetting the association after loading it just to take the loaded records is loose and leads to endless non-cases. I think I can invent failing tests forever :)

My main motivation for this change is this functionality: https://github.com/rails/rails/pull/32136
I consider this patch absolutely required to fix all the bugs described in: https://github.com/rails/rails/pull/32136#issuecomment-461384766

However, I didn't want to merge this patch into `load_associations` functionality because it kind of plays well on its own. Also the final version of `load_associations` will require even more changes for all bugs to be fixed which will make it hard to review.

cc @rafaelfranca @kamipo

### CODE REVIEW

File: activerecord/test/cases/associations/nested_through_associations_test.rb [hunk 1]
```diff
@@ -548,6 +548,15 @@ def test_nested_has_many_through_with_conditions_on_source_associations_preload
     end
   end
 
+  def test_through_association_preload_doesnt_reset_source_association_if_already_preloaded
+    blue = tags(:blue)
+    authors = Author.preload(:posts => :first_blue_tags_2, :misc_post_first_blue_tags_2 => {}).to_a.sort_by(&:id)
```
Comment (kamipo) [inline activerecord/test/cases/associations/nested_through_associations_test.rb]:
```suggestion
    authors = Author.preload(posts: :first_blue_tags_2, misc_post_first_blue_tags_2: {}).to_a.sort_by(&:id)
```

File: activerecord/lib/active_record/associations/preloader/through_association.rb [hunk 2]
```diff
@@ -4,45 +4,57 @@ module ActiveRecord
...
+          @preloaded_records ||= source_preloaders.flat_map(&:preloaded_records)
+        end
+
+        def records_by_owner
+          return @records_by_owner if defined?(@records_by_owner)
+          source_records_by_owner = source_preloaders.map(&:records_by_owner).reduce(:merge)
+          through_records_by_owner = through_preloaders.map(&:records_by_owner).reduce(:merge)
+
+          @records_by_owner = owners.each_with_object({}) do |owner, result|
+            middle_records = through_records_by_owner[owner] || []
```
Comment (kamipo) [inline activerecord/lib/active_record/associations/preloader/through_association.rb]: Why changed var naming from `through_records` to `middle_records`?

File: activerecord/lib/active_record/associations/preloader/association.rb [hunk 3]
```diff
@@ -54,21 +75,15 @@ def owner_keys
...
           end
 
           def key_conversion_required?
-            unless defined?(@key_conversion_required)
-              @key_conversion_required = (association_key_type != owner_key_type)
-            end
-
-            @key_conversion_required
+            return @key_conversion_required if defined?(@key_conversion_required)
+            @key_conversion_required = (association_key_type != owner_key_type)
```
Comment (kamipo) [inline activerecord/lib/active_record/associations/preloader/association.rb]: Is there any reason to touch the code? This is completely the same logic as before.

File: activerecord/lib/active_record/associations/preloader/association.rb [hunk 4]
```diff
@@ -4,26 +4,47 @@ module ActiveRecord
...
+            preload
+          else
+            # Custom preload scope is used and
+            # the association can not be marked as loaded
+            # Loading into a Hash instead
+            records_by_owner
           end
+        end
 
+        def preload
```
Comment (kamipo) [inline activerecord/lib/active_record/associations/preloader/association.rb]: Looks like this is called from `run` only, should be private?

**Review by kamipo — COMMENTED** [review]
Nice improvement.

File: activerecord/lib/active_record/associations/preloader/through_association.rb (hunk 2 above)
Comment (bogdan) [reply activerecord/lib/active_record/associations/preloader/through_association.rb]: `through_records` collides with the method name. `middle_records` is just to make it different from the method name. Better naming is welcome.

Comment (kamipo) [reply activerecord/lib/active_record/associations/preloader/through_association.rb]: How about flipping the `through_records` and `middle_records` like as before?

File: activerecord/lib/active_record/associations/preloader/association.rb [hunk 5]
```diff
@@ -4,26 +4,44 @@ module ActiveRecord
...
+        def run
+          if !preload_scope || preload_scope.empty_scope?
+            owners.each do |owner|
+              associate_records_to_owner(owner, records_by_owner[owner] || [])
+            end
+          else
+            # Custom preload scope is used and
+            # the association can not be marked as loaded
+            # Loading into a Hash instead
+            records_by_owner
```
Comment (kamipo) [inline activerecord/lib/active_record/associations/preloader/association.rb]: Is this explict call necessary?

Comment (bogdan) [reply activerecord/lib/active_record/associations/preloader/association.rb]: No, the only motivation for it was that lazy execution of the database query can change queries order and make it hard to control. However, none of the tests is failing when I comment it out. Do you think it is better to remove it?

Comment (kamipo) [reply activerecord/lib/active_record/associations/preloader/association.rb]: My preference is keeping accessors as before (`preloaded_records`, and add new `records_by_owner`), and extracting building method. What do you think?

```ruby
        def run
          records_by_owner = load_records # or build_records_by_owner or something else

          if !preload_scope || preload_scope.empty_scope?
            owners.each do |owner|
              associate_records_to_owner(owner, records_by_owner[owner] || [])
            end
          end

          self
        end

        def load_records
          @preloaded_records = ...
          @records_by_owner = ...
        end
```

Comment (bogdan) [reply activerecord/lib/active_record/associations/preloader/association.rb]: I have a bad experience with this pattern where a lot of ivars are assigned within the same method. It makes it hard to maintain - in case we would need to just access `preloaded_records` but not anything else in the future, and hard to follow: with separated methods implementing the data loading the new person looking at the code can trace the dependencies and execution sequence easily, while in your proposal ivar usage has indirect link to var assignment. It can become even worse: this implementation leads to the same ivar  assigning in multiple places (unless someone supervises it precisely which is very unlikely for rails repo).

In my projects, I broke down multiple ivars assignment in the same method into multiple short methods with cache with clear execution chain and ability to use one by one many times. However, according to current situation, your proposal looks ok. So, I'll leave it up to you. Please just let me know.

Comment (kamipo) [reply activerecord/lib/active_record/associations/preloader/association.rb]: Okey, I'd leave it as is for now, we could refactor around preloader in any future round.

**Comment by ghiculescu** [comment]
@bogdan are you able to confirm if https://github.com/rails/rails/issues/37720 is a regression, or expected behavior?

**Comment by bogdan** [comment]
@ghiculescu you can my opinion on the subject here: https://github.com/rails/rails/issues/36638#issuecomment-510094259

Core team didn't react to the problem so far. Maybe if you give it more :+1:, they will.

**Comment by Edouard-chin** [comment]
👋 

I encountered another issue related to this change. I tried to find a fix but had a hard time wrapping my head around the problem that this PR tries to solve.

Here is a small reproduction script https://gist.github.com/Edouard-chin/70744fa2debac416feab352a9ccc2bd1
TL;DR on the issue is that adding a `default_scope` on a `through` relation makes preloading to not work.

--------------------- 
Also small heads up that the test added in this PR is never failing whether a preload scope exists or not
```diff
diff --git a/activerecord/lib/active_record/associations/preloader/association.rb b/activerecord/lib/active_record/associations/preloader/association.rb
index 4c7b0e6f07..f7f791cddb 100644
--- a/activerecord/lib/active_record/associations/preloader/association.rb
+++ b/activerecord/lib/active_record/associations/preloader/association.rb
@@ -13,7 +13,7 @@ def initialize(klass, owners, reflection, preload_scope)
         end

         def run
-          if !preload_scope || preload_scope.empty_scope?
+          if true
             owners.each do |owner|
               associate_records_to_owner(owner, records_by_owner[owner] || [])
             end
```
```
bin/test test/cases/associations/nested_through_associations_test.rb -n test_through_association_preload_doesnt_reset_source_association_if_already_preloaded
Using sqlite3
Run options: -n test_through_association_preload_doesnt_reset_source_association_if_already_preloaded --seed 31526

# Running:

.
Finished in 0.124261s, 8.0476 runs/s, 16.0952 assertions/s.
1 runs, 2 assertions, 0 failures, 0 errors, 0 skips
````

**Comment by yosiat** [comment]
Opened a bug for this - https://github.com/rails/rails/issues/38419
